monai/data/dataset.py

```diff
@@ -30,33 +34,116 @@ class Dataset(torch.utils.data.Dataset):
          },                           },                           }]
     """
 
-    def __init__(self, data, transform=None):
+    def __init__(self, data, transform=None, cache_dir=None):
```

Comment:
Perhaps we should call it "persistance_dir" as this code change is really about persistence of the precomputed values.  

The CacheDataset is really a "pre-load".  My changes are focused on computing the results once, and persisting those intermediate results to disk.  The base Dataset "reloads" (either from cache without processing, or from original data with preprocessing ) the data and allows for larger than memory data sets to be used. The "CacheDataset" requires that all the data can be loaded into memory at once.

Dataset with persistence is really important for my larger than memory data set.
CacheDatase with persistence is also important if the entire dataset fits in memory.

These two concepts ar

Reply:
NOTE:  This solution also allows for computing the persistence to disk "on-the-fly" on an "as-needed" basis.  As before, only the data in the current batch is actually loaded in memory and discarded as when the next batch is encountered.  The first time new data is encountered, there is a bit of an overhead to also (optionally) write to the persistence cache, the second time data is encountered, it is reloaded from persistence cache, and not transform processing is done.

The original functionality of the Dataset object is exactly as it used to be if cache_dir=None.

Reply:
Thanks for your detailed explanation.
I also internally discussed with others, if you don't want to combine with CacheDataset, I think you can create a new class inheriting from Dataset directly, maybe `PersistDataset`?Your idea is really awesome and let's try to make it an independent module.
Thanks.

Reply:
I will add a new PersistDataset if that is what is desired.  I'll have an updated PR in a few hours.

Reply:
Cool, `Dataset` is the base class for all kinds of Dataset, let's try to keep it simple and extensible, especially if PersistDataset has some requirements on the input data type.
Looking forward to your update!
This is an important and big feature, let's discuss deeper and make it robust enough.
Thanks in advance.

Reply:
Hi,

I am thinking about whether we should combine `PersistDataset` and `CacheDataset`:
1. If we change to pre-process all data and save to persisting storage before the training instead of on-the-fly in every iteration, the logic is very similar to CacheDataset. Just if providing `cache_dir`, saving to disk, otherwise in memory(loading is the same).
2. People may want to pre-process 100 images and save to persisting storage then only cache 20(cache_rate=0.2) results in memory due to memory size limit. It's a "hybrid" usage of PersistDataset and CacheDataset.

What do you think?
Thanks.

Reply:
I think that pre-processing all data is not a good idea.  It takes a long time to do pre-processing and getting to real processing seems like an important priority.

Reply:
OK, let's keep your "on-the-fly" cache method for current version before we see any other benefit. Thanks.

Reply:
Done in latest revision.